Add experimental riscv64 builds

kxxt opened the pull request:
With a performance issue(https://github.com/riscv-forks/electron/issues/1) fixed, I think the riscv64 port of electron is mostly ready for VSCodium. There are still some corner cases that lead to segfault but it's already suitable for usage on real riscv64 hardware.

This PR add riscv64 builds to vscodium.

Official electron doesn't support riscv64 so I used a fork maintained by myself.

BTW I noticed some of the patches modifies the compiled javascript files. I think it's cleaner to directly modify the typescript source file and compile ts files to javascrpt.

I haven't implemented riscv64 support in sysroot scripts so this PR only covers a basic riscv64 build, without DEB and RPM targets.

Here's a screenshot:

VSCodium 1.90 running on 64 core SG2042.

The window on the right shows using open-remote-ssh plugin to connect to x86_64 host from riscv64.

![Screenshot_20_10_1](https://github.com/VSCodium/vscodium/assets/18085551/cd25086b-8b70-45c2-a8b6-d7baedd71ac1)

lex-ibm commented:
Ups... I think we were working on more or less the same...

lex-ibm commented:
Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

daiyam commented:
> Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

We should split the `ppc64le-and-riscv64-support.patch`.

daiyam commented on build.sh:
```diff
@@ -14,6 +14,11 @@ if [[ "${SHOULD_BUILD}" == "yes" ]]; then
+  # Some patches(e.g. remove-mangle.patch) causes unused variables and parameters
+  sed -ri '/noUnused(Local|Parameter)s/d' build/tsconfig.json
+  # Recompile ts files in ./build since the patches modifies them
+  yarn --cwd build compile
```
No needed since the patches also modify the `.js`

kxxt replied:
Done

daiyam commented on patches/remove-mangle.patch:
```diff
@@ -1,37 +1,10 @@
-diff --git a/build/lib/compilation.js b/build/lib/compilation.js
```
Do not modify that patch.

daiyam commented on patches/ppc64le-and-riscv64-support.patch:
```diff
@@ -372,10 +393,10 @@ index 4bec13d6e86..3debda1224a 100644
  			Platform::WindowsX64 => "WindowsX64",
 diff --git a/cli/src/util/prereqs.rs b/cli/src/util/prereqs.rs
-index 20a5bc94b37..50678354245 100644
+index 5d4953acbb2..f22ff60e391 100644
```
Not needed

daiyam commented on patches/ppc64le-and-riscv64-support.patch:
```diff
@@ -409,10 +430,10 @@ index 079557869e3..c057746b870 100644
  if [ "$OS_ID" != "alpine" ]; then
 diff --git a/src/vs/platform/extensionManagement/common/extensionManagement.ts b/src/vs/platform/extensionManagement/common/extensionManagement.ts
-index 9dae82eba07..94246285b26 100644
+index 183f2582871..2e6614cad59 100644
```
Not needed

lex-ibm commented:
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> We should split the `ppc64le-and-riscv64-support.patch`.

Roger that!

daiyam commented on patches/linux/use-node16.patch:
```diff
@@ -1,19 +1,27 @@
+ import * as ansiColors from 'ansi-colors';
+ import * as crypto from 'crypto';
+ import * as through2 from 'through2';
++const fetch: any = require('node-fetch');
```
Please revert the changes on the file

daiyam commented on patches/fix-build-linux.patch:
```diff
@@ -18,20 +18,6 @@ index 8c2b62f..843e8d4 100644
 -	const buildRpmTask = task.define(`vscode-linux-${arch}-build-rpm`, buildRpmPackage(arch));
 +	const buildRpmTask = task.define(`vscode-linux-${arch}-build-rpm`, task.series(prepareRpmTask, buildRpmPackage(arch)));
  	gulp.task(buildRpmTask);
-diff --git a/build/linux/dependencies-generator.js b/build/linux/dependencies-generator.js
```
Please revert the changes on the file

daiyam commented on patches/ext-from-gh.patch:
```diff
@@ -1,22 +1,15 @@
-diff --git a/build/lib/builtInExtensions.js b/build/lib/builtInExtensions.js
-index 1b0adc4..b595123 100644
---- a/build/lib/builtInExtensions.js
-+++ b/build/lib/builtInExtensions.js
```
Please revert the changes on the file

kxxt commented:
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> We should split the `ppc64le-and-riscv64-support.patch`.

I think it might be cleaner to keep one patch. If we split the patch, the ppc64 and riscv64 patches both modifies the same line in a file, which will lead to a conflict. WDYT?

kxxt commented:
> Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?

Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?

lex-ibm commented:
> > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> 
> Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?

We don't publish the binaries, but we use the `ELECTRON_MIRROR` variable to set the location for it. The PR that I opened only makes the changes needed to enable the build to happen. Is up to the user to specify where that electron binary comes from.

kxxt commented:
> > > Should we coordinate and have this be a single PR for both RISCV64 and PPC64LE?
> > 
> > 
> > Does the PPC64LE has an electron fork that publishes electron binaries? If so I could add it in this PR?
> 
> We don't publish the binaries, but we use the `ELECTRON_MIRROR` variable to set the location for it. The PR that I opened only makes the changes needed to enable the build to happen. Is up to the user to specify where that electron binary comes from.

Oh sorry, I didn't see your PR. I thought you haven't opened it. About the patches, do you think it's more convenient to split riscv64 and ppc64 or keep one patch?  I think splitting the patch doesn't actually solve the conflict but makes it more complex because we need to stack a patch on top of another.

lex-ibm commented:
> I think it might be cleaner to keep one patch. If we split the patch, the ppc64 and riscv64 patches both modify the same line in a file, which will lead to a conflict. WDYT?

I kinda agree, I tried splitting the patch and the only way I got it to work was if I based one on top of the other. Also, in the other PR there are some changes to how to get the vscode-sysroot that make it so that we don't need to check for a specific arch.

While developing it I was trying to make it so that porting to RISCV64 was more or less easy, but never tried to build it.

lex-ibm commented:
>  think splitting the patch doesn't actually solve the conflict but makes it more complex because we need to stack a patch on top of another.

Agree, took me a couple of hours just to split the changes and always diffing everything to make sure I didn't mess anything up. I kinda feel like this could be a single PR or at least we should both coordinate on how to get them both in.

I do recommend you take the patch to `install-sysroot.js`, that makes it easier if the tag changes or we want to use another repo.

kxxt commented:
> I do recommend you take the patch to `install-sysroot.js`, that makes it easier if the tag changes or we want to use another repo.

I agree with you. There's no need to maintain two methods of using a custom electron repo.

kxxt commented:
Another way is to decommonize the intermediate vscode artifact, that is, building an intermediate vscode artifact for each architecture. This way we can apply different patches to different architectures. But I don't think it worth the effort for just two architectures that need patching. This method might make more sense if we have patches for four or more architectures that need patching.

lex-ibm commented:
> This method might make more sense if we have patches for four or more architectures that need patching.

As long as no one asks me to build for `s390x` I'm fine with that.

Let's get started then! I'm in `GMT-5`, so this is probably the best time to contact me.

I see that you replace the `electron.txt` checksums. That would collide with my needs, since that checksums file wouldn't have the sha256 for our binary.

daiyam commented on package_linux_bin.sh:
```diff
@@ -16,6 +16,21 @@ cd vscode || { echo "'vscode' dir not found"; exit 1; }
+  # RISC-V release to the upstream release tag(v<version>). 
+  # This might be problematic if there is an old electron zip cached locally.
+  rm -f ~/.cache/electron/*/electron-*-linux-riscv64.zip
+  ELECTRON_VERSION="v$(yarn config get target)"
```
It would be better to have specific version number and its checksum.
This way, we know for sure which electron version has been used for a specific VSCodium version.
It allows us to be able to reproduce that version if needed.

kxxt replied:
> It would be better to have specific version number and its checksum. This way, we know for sure which electron version has been used for a specific VSCodium version. It allows us to be able to reproduce that version if needed.

Yes, I agree with you. I could use a specific version like v29.4.2.riscv2 but it would require me to make more PRs in the future to maintain the riscv64 revision so I didn't take this approach on the first try.

lex-ibm commented on patches/ppc64le-and-riscv64-support.patch:
```diff
@@ -97,10 +97,29 @@ index 6f8144b0954..8a8195e9b66 100644
++		const electronOverride = {};
++		if (platform === 'linux' && arch === 'riscv64') {
++			// electron doesn't support riscv64, so a fork is used here.
++			electronOverride.repo = 'riscv-forks/electron-riscv-releases';
```
This can be achieved with `ELECTRON_MIRROR`, can it not?

kxxt replied:
Yes.

daiyam replied:
Not sure. `ELECTRON_MIRROR` seems to be a `electron-builder` variable.
VSCode seems to use `electronRepository` in the `product.json`

https://github.com/microsoft/vscode/blob/0fd9a006f7abcb48ef1e5a9f9ceeca33e1d1daaa/build/lib/electron.ts#L201

kxxt replied:
> This can be achieved with `ELECTRON_MIRROR`, can it not?

@lex-ibm I didn't see `ELECTRON_MIRROR` anywhere in your PR? Where did you implemented it in your PR?

kxxt replied:
> This can be achieved with `ELECTRON_MIRROR`, can it not?

@lex-ibm I didn't see `ELECTRON_MIRROR` anywhere in your PR? Where did you implemented it in your PR?

kxxt replied:
Oh the network glitch caused me to send the comment twice. Sorry.

kxxt replied:
Resolved by using a `VSCODE_ELECTRON_REPO` environment variable for the electron repo.

kxxt commented:
> > This method might make more sense if we have patches for four or more architectures that need patching.
> 
> As long as no one asks me to build for `s390x` I'm fine with that.
> 
> Let's get started then! I'm in `GMT-5`, so this is probably the best time to contact me.
> 
> I see that you replace the `electron.txt` checksums. That would collide with my needs, since that checksums file wouldn't have the sha256 for our binary.

It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.

daiyam commented:
> It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.

It will be better if you could just patch the `electron.txt`

daiyam commented:
> > It wouldn't collide with your needs. The riscv64 and ppc64le pipelines are ran separately so replacing it in riscv64 shouldn't affect it's usage in ppc64le.
> 
> Is it ok with you if the binaries for the electron binary and it's sha256 change without VSCodium knowing? I kinda feel like that might be a bit risky.

I agree

daiyam commented:
> I actually don't replace the checksums but append to it. @daiyam Does it look good to you? I consider it as a method of "patch".

No, I prefer to have version and checksum in clear text in the repo. There are several benefit to do this:
- every body can look at the code and know which version has been used
- reproducible binaries
- the changes of version are tracked

We do the same for VSCode ([stable.json](https://github.com/VSCodium/vscodium/blob/master/stable.json))

lex-ibm commented:
I don't know if it helps, but I do remember seeing an environment variable to change the electron version, I just don't remember where. I'll look for it.
With that I think you could use `ELECTRON_MIRROR` and whatever the other variable is and all that would be needed is the patch for `electron.txt`.

kxxt commented:
> > I actually don't replace the checksums but append to it. @daiyam Does it look good to you? I consider it as a method of "patch".
> 
> No, I prefer to have version and checksum in clear text in the repo. There are several benefit to do this:
> 
>     * every body can look at the code and know which version has been used
> 
>     * reproducible binaries
> 
>     * the changes of version are tracked
> 
> 
> We do the same for VSCode ([stable.json](https://github.com/VSCodium/vscodium/blob/master/stable.json))

Thanks for clarifying. I have squashed the commits and rebased my PR. Will mark ready for review once the CI passes here: https://github.com/kxxt/vscodium/actions/runs/9468203671

lex-ibm commented:
[Found it!](https://github.com/electron/get/blob/main/src/artifact-utils.ts) Can even specify a custom file name! Does that help?

daiyam commented:
> @daiyam I write the clear text electron checksum in `package_linux_bin.sh`. Do you think I need to create a patch file for it or it's fine to maintain the checksum in `package_linux_bin.sh`?

You don't need a patch

kxxt commented:
> > @daiyam I write the clear text electron checksum in `package_linux_bin.sh`. Do you think I need to create a patch file for it or it's fine to maintain the checksum in `package_linux_bin.sh`?
> 
> You don't need a patch

Then I think the PR is ready for review now. The CI is green here(for riscv64 and ppc64le): https://github.com/kxxt/vscodium/actions/runs/9468203671/job/26084811222

daiyam commented:
> [Found it!](https://github.com/electron/get/blob/main/src/artifact-utils.ts) Can even specify a custom file name! Does that help?

@lex-ibm VSCode seems to use the package [@vscode/gulp-electron](https://www.npmjs.com/package/@vscode/gulp-electron) to download electron (https://github.com/microsoft/vscode/blob/5adc4dcb025aac567c55100c4f2777614644de89/build/lib/electron.ts#L208)

daiyam commented on package_linux_bin.sh:
```diff
@@ -22,6 +22,17 @@ if [[ "${VSCODE_ARCH}" == "ppc64le" ]]; then
+  export VSCODE_ELECTRON_REPO='riscv-forks/electron-riscv-releases'
+  export ELECTRON_SKIP_BINARY_DOWNLOAD=1
+  export PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1
+  ELECTRON_VERSION="v$(yarn config get target)"
```
`ELECTRON_VERSION="v29.4.0"` should be enough

kxxt replied:
Fixed

daiyam commented on patches/ppc64le-and-riscv64-support.patch:
```diff
@@ -97,10 +97,33 @@ index 6f8144b0954..8a8195e9b66 100644
++		const electronOverride = {};
++		if (process.env.VSCODE_ELECTRON_REPO) {
++			// electron doesn't support riscv64 and ppc64le, override the repo with `VSCODE_ELECTRON_REPO`.
++			electronOverride.repo = process.env.VSCODE_ELECTRON_REPO;
```
Why not directly modifying the `config` variable?

daiyam commented:
LGTM

@kxxt Is it ok for you?

daiyam commented:
@lex-ibm Do you have any feedback?

kxxt commented:
> LGTM
> 
> @kxxt Is it ok for you?

Yes

lex-ibm commented:
LGTM! Awesome job!

daiyam commented:
@kxxt Thanks!

kxxt commented:
Thanks both of you for helping me to get this PR merged!

About automatically updating the riscv64 electron version and checksum. I think we can implement it in a script and run it in the *-spearhead workflows.
